snap/local/runtime-helpers/bin/kong-postgres-setup.sh

```diff
@@ -0,0 +1,281 @@
+
+    # ensure the postgres data directory exists and is owned by snap_daemon
+    mkdir -p "$POSTGRES_12_DB_DIR"
+    chown -R snap_daemon:snap_daemon "$SNAP_DATA/postgresql" 
+    
+    # ensure the sockets dir exists and is owned by snap_daemon
+    mkdir -p "$SNAP_COMMON/sockets"
+    chown -R snap_daemon:snap_daemon "$SNAP_COMMON/sockets"
+
+    # create the directory used for the kong backup
+    mkdir -p "$SNAP_COMMON/refresh"
+    chown -R snap_daemon:snap_daemon "$SNAP_COMMON/refresh"
+
+    # create the config directory used for the kong password
+    mkdir -p $CONFIG_KONG_PASSWORD_DIR 
+
+    # remove pg 10 directories
+    rm -rf "$SNAP_DATA/etc/postgresql"
+    
+    # remove database directory
+    rm -rf "$SNAP_DATA/postgresql/10/main"
+}
+
+
+
+postgres_create_configuration_file() {
+
+    logger "edgex-kong-postgres-setup: creating configuration file $POSTGRES_12_DB_DIR/$POSTGRES_CONFIGFILENAME"
+    
+   snap_daemon_run_command cp "$SNAP/config/postgres/$POSTGRES_CONFIGFILENAME" "$POSTGRES_12_DB_DIR/"
+
+    # do replacement of the $SNAP, $SNAP_DATA, $SNAP_COMMON environment variables in the config files
+     snap_daemon_run_command sed -i -e "s@\$SNAP_COMMON@$SNAP_COMMON@g" \
+        -e "s@\$SNAP_DATA@$CURRENT@g" \
+        -e "s@\$SNAP@$SNAP_CURRENT@g" \
+        "$POSTGRES_12_DB_DIR/$POSTGRES_CONFIGFILENAME" 
+    
+}
+ 
+
+# called from the security-proxy-post-setup script.
+# kong needs to be running for this to succeed, so we can't run this in the 
+# post-refresh hook
+postgres_restore_backup() {
+    if [ -f "$SNAP_COMMON/refresh/kong.sql" ]; then
+      logger "edgex-kong-postgres-setup: restoring kong database from $SNAP_COMMON/refresh/kong.sql"
+      snap_daemon_run_command "$PSQL_BIN_PATH/psql" "-Ukong" "-f$SNAP_COMMON/refresh/kong.sql" "kong"
+    else
+        logger "edgex-kong-postgres-setup: No backup found at $SNAP_COMMON/refresh/kong.sql"
+    fi
+}
+
+
+# This function is called either:
+# a) from install hook (via the install-setup-postgres.sh script), to set up a new PG 12 instance
+# b) from the post-refresh hook, to refresh the PG instance from an older PG version
+postgres_install() {
+
+    logger "edgex-kong-postgres-setup: install postgres 12"
+  
+    # Note: It's fine to replace the current postgres installation.
+    # Postgresql is only used for kong and if we're doing a snap refresh, then
+    # we will backup and restore the Kong configuration. 
+
+    # create the required directories 
+    postgres_setup_directories
+
+    postgres_create_password_file
+    
+    # set up the new postgres v12 database
+    postgres_init_postgres_database
+
+    # and the config file
+    postgres_create_configuration_file
+
+    # start postgres up and wait a bit for it so we can create the database and user for kong
+    logger "edgex-kong-postgres-setup: install - starting postgres"
+    snapctl start "$SNAP_NAME.postgres"
+ 
+    postgres_create_kong_db
```

Comment:
The waiting now happens in another function.
```suggestion
    # start postgres and create the database and user for kong
    logger "edgex-kong-postgres-setup: install - starting postgres"
    snapctl start "$SNAP_NAME.postgres"
 
    postgres_create_kong_db
```